### PR TITLE
feat(core): support github application

### LINKED ISSUE #351: [feat] use GitHub apps
It would be great if we could use Abstruse as a GitHub application instead of providing user access tokens. Abstruse would simply need to accept a private key used to sign it's own JWT for api requests. Here is an example of what the configuration would look like for github enterprise users.

image

### PR DESCRIPTION
**Pull request by matthax**
- Support GitHub enterprise api endpoints
- Add tests to verify GitHub urls are parsed correctly
- Change access token form to support integrations
  - Allow GitHub app to be toggled
  - Allow private key to be uploaded
  - Allow Installation and Integration id to be set
- Change service worker to re-authenticate expired tokens
- Change git process to log to screen
- Add documentation for app integration
- Use correct repository url (as specified in settings) when setting commit statuses

Closes #351

### CODE REVIEW

File: src/api/db/build.ts [hunk 1]
```diff
@@ -119,25 +123,51 @@ export function getBuild(id: number, userId?: number): Promise<any> {
...
           }
-        } else {
-          build.hasPermission = false;
-        }
-
-        return build;
+          if (build.repository.access_token) {
+            const accessToken: AccessTokenType = build.repository.access_token;
+            if (build.repository.access_token.is_integration) {
+              // let msg: LogMessageType = {
```
Comment (jkuri) [inline src/api/db/build.ts]: please remove comments.

**Comment by jkuri** [comment]
Matt, this PR looks good. I will check and run it locally tomorrow in review it in more detail. 
In the meanwhile, can you please solve issues with your account so your commits will be linked with your account, please?

**Comment by matthax** [comment]
I will attempt to rebase it when I remove the comment and amend the author.

**Comment by jkuri** [comment]
can you come on gitter please? I am online now.

**Comment by matthax** [comment]
I will resolve the test issue today.

**Comment by jkuri** [comment]
@matthax sorry for the super-late response from my side. I just reviewed this PR and I must say that it looks super cool. I am currently struggling to rebase this against current `master` branch. 
As soon as I manage to do this this will be merged.
I am wondering if we can add some tests for this, do you have any idea?

**Comment by jkuri** [comment]
Closing in favor of v2

**Pull request closed without merging**